replications/internal/queue_management.go

```diff
@@ -71,44 +81,133 @@ func (qm *durableQueueManager) InitializeQueue(replicationID platform.ID, maxQue
 		return err
 	}
 
-	// Map new durable queue to its corresponding replication stream via replication ID
-	qm.replicationQueues[replicationID] = newQueue
-
 	// Open the new queue
 	if err := newQueue.Open(); err != nil {
 		return err
 	}
 
+	// Map new durable queue and scanner to its corresponding replication stream via replication ID
+	rq := replicationQueue{
+		queue:   newQueue,
+		logger:  qm.logger,
+		done:    make(chan struct{}),
+		receive: make(chan struct{}),
+	}
+	qm.replicationQueues[replicationID] = &rq // todo does this need to be held?
+	rq.Open()
+
 	qm.logger.Debug("Created new durable queue for replication stream",
 		zap.String("id", replicationID.String()), zap.String("path", dir))
 
 	return nil
 }
 
+func (rq *replicationQueue) Open() {
+	rq.wg.Add(1)
+	go rq.run()
+}
+
+func (rq *replicationQueue) Close() error {
+	close(rq.receive)
+	close(rq.done)
+	rq.wg.Wait() // wait for goroutine to finish processing all messages
+	return rq.queue.Close()
+}
+
+func (rq *replicationQueue) run() {
+	defer rq.wg.Done()
+
+	retryInterval := time.Second // todo configurable?
+	retryTimer := time.NewTicker(retryInterval)
+	defer retryTimer.Stop()
+
+	writer := func() {
+		for {
+			_, err := rq.SendWrite(func(b []byte) error {
+				rq.logger.Info("written bytes", zap.String("bytes", string(b)))
+				return nil
+			})
+			if err != nil {
+				if err == io.EOF {
+					// No more data
+					rq.logger.Debug("Finished replication writes for queue")
+				} else {
+					// todo more error handling
+					panic(1)
+				}
+				break
+			}
+		}
+	}
+
+	for {
+		select {
+		case <-rq.done: // end the goroutine when done is messaged
+			return
+		case <-retryTimer.C: // run the scanner every 1s
+			writer()
+		case <-rq.receive: // run the scanner on data append
+			writer()
+		}
+	}
+}
+
+func (rq *replicationQueue) SendWrite(dp func([]byte) error) (int, error) {
+	// err here can be io.EOF, indicating nothing to write
+	scan, err := rq.queue.NewScanner()
+	if err != nil {
+		return 0, err
+	}
+
+	var count int
+	for scan.Next() {
+		if scan.Err() != nil {
+			err = scan.Err()
+			break
+		}
+		if err = dp(scan.Bytes()); err != nil {
+			break
+		}
+		count += len(scan.Bytes())
+	}
+
+	if err != nil { // todo handle "skippable" errors
+		rq.logger.Info("Segment read error.", zap.Error(scan.Err()))
+	}
+
+	// This may return io.EOF to indicate an empty queue
+	if _, err := scan.Advance(); err != nil {
```

Comment:
It looks like the `Advance` logic doesn't return an EOF from what I'm seeing: https://github.com/influxdata/influxdb/blob/fa9ba8e86f4f21344cd1edd66fc8bed07331eb10/pkg/durablequeue/scanner.go#L41